lib/ansible/modules/database/influxdb/influxdb_user.py

```diff
@@ -83,6 +83,12 @@
 from ansible.module_utils.basic import AnsibleModule
 from ansible.module_utils.influxdb import InfluxDb
 
+try:
```

Comment:
this block is already in module_utils i don't think it is necessary to duplicate code.

Reply:
@zhhuta, Thank you for the review!

As you see, I need to [catch InfluxDBClientError](https://github.com/ansible/ansible/pull/35471/files/19b72319f0a065d60e1022cf9ced676aea74cb81#diff-74b96b08f0cd01016bdd3fab3bb08467R113) in the check_user_password function.
This error defined in the [influxdb-python module](https://github.com/influxdata/influxdb-python/blob/795907d7b3f0ddb6f8d07a8b54ce2e1777e87010/influxdb/exceptions.py#L10). That's why I need to add import line:
```python
from influxdb.exceptions import InfluxDBClientError
```

If I add only the `import` line, then test `ansible-test sanity --test import --python XX` will fail with the error:
```
lib/ansible/modules/database/influxdb/influxdb_user.py:85:0: ImportError: No module named influxdb..exceptions
```
I've already [tried](https://github.com/ansible/ansible/pull/35471#issuecomment-361341320). :-)

It is because of [this check](https://docs.ansible.com/ansible/devel/dev_guide/testing/sanity/import.html):
>All Python imports in `lib/ansible/modules/` and `lib/ansible/module_utils/` which are not from the Python standard library must be imported in a try/except ImportError block.

That's why I'd copied `try...except` from module_utils/influxdb.py:
```python
try:
    from influxdb.exceptions import InfluxDBClientError
    HAS_INFLUXDB = True
except ImportError:
    HAS_INFLUXDB = False
```

I think I can replace this block with
```python
try:
    from influxdb.exceptions import InfluxDBClientError
except ImportError:
    pass
```
And then I can drop:
```python
if not HAS_INFLUXDB:
    module.fail_json(msg='This module requires influxdb python package.')
```

What do you think?

Reply:
Lets keep logic striate:  you want  set password if user already exist:
`if state == 'present':
        if user:
            if check_user_password(module, user_name, user_password):
                module.exit_json(changed=False)
            else:
                set_user_password(module, client, user_name, user_password)
        else:
            create_user(module, client, user_name, user_password, admin)`

it's ok but lets look on logic above, you already have 
`influxdb = InfluxDb(module)
  client = influxdb.connect_to_influxdb()`
and when you call `check_user_password` you create another connection to db inside of this function.
`def check_user_password(module, user_name, user_password):
    influxdb = InfluxDb(module)
    client = influxdb.connect_to_influxdb()`
Obviously you can pass to this function client object that was init in main()
Next 
`try:
        client.switch_user(user_name, user_password)
        client.get_list_users()
    except InfluxDBClientError as e:
        if e.code == 401:
            return False`
why you do `client.get_list_users()` it returns a list of users : actually a list-object. i guess you can keep only `client.switch_user(user_name, user_password)`
Another is about `InfluxDBClientError` its already part of InfluxDBClient that we import from influxdb in module_utils/influxdb.py and if you go deeper in `influxdb-python/influxdb/client.py`
Any way let me try to work with your code and figure out how we can make this code more clean

Reply:
take a look on drop_user()
```
 try:
            client.drop_user(user_name)
        except client.InfluxDBClientError as e:
            module.fail_json(msg=e.content)
```

Reply:
1. Import
    > take a look on drop_user()
    ```python
    try:
        client.drop_user(user_name)
    except client.InfluxDBClientError as e:
        module.fail_json(msg=e.content)
    ```
    @zhhuta Are you sure that this code works? I'm not a python expert, sorry if I say something stupid.
    I've tried to cause `InfluxDBClientError` in `drop_user`. I've replaced it by:
    ```python
    def drop_user(module, client, user_name):
        if not module.check_mode:
            try:
                client.switch_user('test', 'test')
                client.drop_user(user_name)
            except client.InfluxDBClientError as e:
                module.fail_json(msg=e.content)

        module.exit_json(changed=True)
    ```
    As you can see, I've added `client.switch_user('test', 'test')`. This line always causes `InfluxDBClientError` with code 401 (Unauthorized) (there is no 'test' user in my InfluxDB).
    Then I've run the following test:
    ```yaml
    - name: Add user
      influxdb_user: user_name=user user_password=password login_username=admin login_password=admin

    - name: Remove user
      influxdb_user: user_name=user state=absent login_username=admin login_password=admin
    ```
    It failed with the error `AttributeError: 'InfluxDBClient' object has no attribute 'InfluxDBClientError'`:
    ```
    An exception occurred during task execution. To see the full traceback, use -vvv. The error was: AttributeError: 'InfluxDBClient' object has no attribute 'InfluxDBClientError'
    fatal: [testhost]: FAILED! => {"changed": false, "failed": true, "module_stderr": "Traceback (most recent call last):\n  File \"/tmp/ansible_sSZKDE/ansible_module_influxdb_user.py\", line 195, in <module>\n    main()\n  File \"/tmp/ansible_sSZKDE/ansible_module_influxdb_user.py\", line 189, in main\n    drop_user(module, client, user_name)\n  File \"/tmp/ansible_sSZKDE/ansible_module_influxdb_user.py\", line 147, in drop_user\n    except client.InfluxDBClientError as e:\nAttributeError: 'InfluxDBClient' object has no attribute 'InfluxDBClientError'\n", "module_stdout": "", "msg": "MODULE FAILURE", "rc": 1}
    ```

    When I've added import:
    ```
    from influxdb.exceptions import InfluxDBClientError
    ```
    and replaced `client.InfluxDBClientError` to `InfluxDBClientError` in drop_user function:
    ```python
    def drop_user(module, client, user_name):
        if not module.check_mode:
            try:
                client.switch_user('test', 'test')
                client.drop_user(user_name)
            except InfluxDBClientError as e:
                module.fail_json(msg=e.content)

        module.exit_json(changed=True)
    ```
    the test have failed as expected:
    ```
    fatal: [testhost]: FAILED! => {"changed": false, "failed": true, "msg": "{\"error\":\"authorization failed\"}\n"}
    ```

    That's why I think that we need to explicitly add import.

2. Pass client from `main`
    >and when you call check_user_password you create another connection to db inside of this function.
    ```
    influxdb = InfluxDb(module)
    client = influxdb.connect_to_influxdb()
    ```
    >Obviously you can pass to this function client object that was init in main()

    As you know, there are two users in the module:
    1. `username` + `password` (with aliases `login_username` + `login_password`). This user is used to authenticate against InfluxDB server.
    2. `user_name` + `user_password`. This is user to modify.

    When `main` calls your version of `check_user_password`:
    ```python
    def check_user_password(module, client, user_name, user_password):
        try:
            client.switch_user(user_name, user_password)
            client.get_list_users()
        except client.InfluxDBClientError as e:
            if e.code == 401:
                return False
        except ansible.module_utils.urls.ConnectionError as e:
            module.fail_json(msg=str(e))
        return True
    ```
    it replaces `username` and `password` (user for authentication) in `client` by `user_name` and `user_password` (user to modify) and never rollbacks. Thus if `check_user_password` returns `False`, `set_user_password(module, client, user_name, user_password)` call will never succeed.

    But I understand your concern. What do you think if I change `check_user_password` by this:
    ```python
    def check_user_password(module, client, user_name, user_password):
        try:
            client.switch_user(user_name, user_password)
            client.get_list_users()
        except InfluxDBClientError as e:
            if e.code == 401:
                return False
        except ansible.module_utils.urls.ConnectionError as e:
            module.fail_json(msg=str(e))
        finally:
            # restore previous user
            client.switch_user(module.params['username'], module.params['password'])
        return True
    ```

3. Use only `switch_user` call
    >why do you use `client.get_list_users()`? it returns a list of users : actually a list-object. I guess you can keep only `client.switch_user(user_name, user_password)`

    As I've already [written](https://github.com/ansible/ansible/pull/35471#discussion_r164531081), [switch_user](https://github.com/influxdata/influxdb-python/blob/df70f5ce395d5650c14b9f68707220ab451b5ce8/influxdb/client.py#L192) function just sets two properties, it doesn't call server:
```python
    def switch_user(self, username, password):
        """Change the client's username.
        :param username: the username to switch to
        :type username: str
        :param password: the password for the username
        :type password: str
        """
        self._username = username
        self._password = password
```
`get_list_users` is just a random function which calls server.

Reply:
Hi Artem.
I miss switch user logic in influxdb/client.py - you are right and i got the logic - i agree on suggested change in 2 and it make sense to keep switch_user.
about InfluxDBClientError i guess i need to dive more deep and get understanding thy dependency doesn't work as it import in influxdb/client.py 
``` 
from .exceptions import InfluxDBClientError
from .exceptions import InfluxDBServerError```

Reply:
Actually i've figured out and code need to be refactored in next way 
```#from ansible.module_utils.influxdb import InfluxDb
from ansible.module_utils.influxdb import *
```
so you can use exceptions.InfluxDBClientError
but it's not really nice way of import 
it's another way 
```import ansible.module_utils.influxdb as influxdb```
so it will bring next changes:
```argument_spec = influxdb.InfluxDb.influxdb_argument_spec()```

```influxdb = inflxudb.InfluxDb(module) ```

``` except inflxudb.exceptions.InfluxDBClientError```
such refactoring reduce reimport but bring some hard to understand call.